docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,234 @@
+| HostStatusDiscoveringUnbound |  | HostStatusUnbindingPendingUserAction |
+--------------------------------  ----------------------------------------
+              ^                                       |
+              |       ------------------------------  |
+              --------| TransitionTypeRegisterHost |---
+                      ------------------------------
+```
+
+#### Assumptions
+
+This approach only works if the host is currently part of a working cluster
+which the assisted service can access using the kubeconfig generated during
+install.
+
+There will also need to be enough space on the `/boot` partition to download the
+initrd and kernel, but the agent can check for this. Together the initrd and
+kernel occupy about 90MB.
+
+### Risks and Mitigations
+
+It's possible that something could go wrong during the boot process. A user
+that doesn't have access to the host infrastructure wouldn't be able to debug
+or solve this type of problem.
+
+A timeout will be implemented when the service expects the host to reregister
+(in `HostStatusReclaimingPendingReboot`). After this time the host will move to
+the existing `HostStatusUnbindingPendingUserAction` state.
+
+The real issue is if a host reboots back into the node and the pod running the
+agent is the one that registers, not the one running in the new discovery live
+environment.
+
+## Design Details [optional]
+
+### Open Questions
+
+1. How does assisted service differentiate between an agent registering from the
+   discovery live environment from the DaemonSet pod?
+  - Could be a problem if the host reboots into the node OS rather than the
+    discovery boot entry (described in "Risks and Mitigations")
+
+### UI Impact
+
+A UI would need to be developed for this feature, but it won't be used in the
+cloud so it doesn't need a UI there.
+
+### Test Plan
+
+Any existing end to end late binding test could be expanded to include
+returning the agents to the infraEnv after a successful install.
+
+## Drawbacks
+
+This will only work in environments where the service has access to the
+kubernetes API on the cluster. This specifically means that it won't be useful
+in the SaaS.
+
+## Alternatives
+
+1. Install an agent when the cluster is first installed
+  - This would be a full polling approach where the agent is always polling the 
+    service to determine if it should reprovision.
+  - This is likely not worth the resources and clusters don't exist long enough in
+    the cloud for this to work currently.
+2. Implement starting the agent in MAPI and/or CAPI
+  - These components are already the starting point for removing a node from a
+    cluster and would likely be the ones triggering the unbind
+  - In this case the assisted-service wouldn't require the spoke cluster
+    kubeconfig as MAPI/CAPI should already have this access.
+  - This would mean that users not using MAPI/CAPI would not have access to this
```

Comment:
This alternative is predicated on these assumptions:
* If you're using an existing MAPI or CAPI provider (aws, gcp, ...), you don't need this feature at all.
* This alternative would be implemented as a MAPI/CAPI provider (maybe as part of capi-agent).
* boot-it-yourself would still be the way to bring hardware under initial management.

Does that clarify things? I think the statement on this line then basically amounts to "users who choose not to use it won't be using it."

Reply:
Yes, this makes sense assuming we have no issues implementing and shipping this feature as a part of MAPI _and_ CAPI.

I think I should probably split this into two different conversations about MAPI and CAPI and focus more on the difficulties of implementing/delivering this as a part of those components.

I'll update this in a bit.

Reply:
I'll send you info. Anyone else who is interested, just ping me.